### ISSUE TITLE
Activating features for tests/benchmarks

### ISSUE THREAD

**Issue by ctz** [rust-lang/cargo#2911]
Given a library crate, I want to have optional features which are available to integration-level tests and examples so that 'cargo test' tests them.

Here's what I have:

```
[package]
name = "rustls"
version = "0.1.0"
(etc)

[features]
default = []
clientauth = []

[dependencies]
untrusted = "0.2.0"
(etc)

[dev-dependencies]
env_logger = "0.3.3"
(etc)
```

Here's what I've tried so far. Attempt one: add the crate into dev-dependencies with the feature, ie:

```
[dev-dependencies.rustls]
path = "."
version = "0.1.0"
features = ["clientauth"]
```

I kind of expected that to work, but it panics:

```
jbp@debian:~/rustls$ RUST_BACKTRACE=1 cargo test
thread '<main>' panicked at 'assertion failed: my_dependencies.insert(dep.clone())', src/cargo/util/dependency_queue.rs:86
stack backtrace:
   1:     0x558d17c3a2a0 - std::sys::backtrace::tracing::imp::write::h9fb600083204ae7f
   2:     0x558d17c4383b - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::hca543c34f11229ac
   3:     0x558d17c434c3 - std::panicking::default_hook::hc2c969e7453d080c
   4:     0x558d17c28048 - std::panicking::rust_panic_with_hook::hfe203e3083c2b544
   5:     0x558d17701fef - std::panicking::begin_panic::h4ebf9fe884b2415f
   6:     0x558d17802d27 - cargo::ops::cargo_rustc::job_queue::JobQueue::enqueue::h98b20a33b8744ebe
   7:     0x558d177f4faf - cargo::ops::cargo_rustc::compile::h93336dd21aba29d0
   8:     0x558d1777a513 - cargo::ops::cargo_rustc::compile_targets::h7223dffae01d6b9d
   9:     0x558d1776eec9 - cargo::ops::cargo_compile::compile_pkg::h48656ffa9b1541f3
  10:     0x558d1776bd09 - cargo::ops::cargo_compile::compile::h1b43b20047c53d10
  11:     0x558d1785fb0d - cargo::ops::cargo_test::compile_tests::h028a22d3131e0d65
  12:     0x558d1785f379 - cargo::ops::cargo_test::run_tests::h0ad98c54a6f40c9d
  13:     0x558d176de256 - cargo::call_main_without_stdin::hd484f08b17c419d1
  14:     0x558d1768afe7 - cargo::execute::hab7accd6bf9b5c64
  15:     0x558d17683f8b - cargo::call_main_without_stdin::hf099bd36acb849a3
  16:     0x558d17680da1 - cargo::main::h2b219a79f378c1ef
  17:     0x558d17c430d8 - std::panicking::try::call::hc5e1f5b484ec7f0e
  18:     0x558d17c4e0eb - __rust_try
  19:     0x558d17c4e08e - __rust_maybe_catch_panic
  20:     0x558d17c42b03 - std::rt::lang_start::h61f4934e780b4dfc
  21:     0x7f79ea9bf86f - __libc_start_main
  22:     0x558d17680728 - <unknown>
```

(version: "cargo 0.11.0-nightly (259324c 2016-05-20)")

Attempt two: see if `features.*` works like `profile.*`:

```
[features.test]
default = ["clientauth"]
```

alas

```
$ cargo test
error: failed to parse manifest at `/home/jbp/rustls/Cargo.toml`

Caused by:
  expected a value of type `array`, but found a value of type `table` for the key `features.test`
```

Is this possible?

**Comment by alexcrichton** [comment]
Unfortunately, no, it's not possible for `cargo test` to automatically enable features currently. The panic though definitely seems bad!

**Comment by jjpe** [comment]
I just ran into this issue because I'm currently using a feature to decide whether or not to do some rather verbose tracing. I'd like to be able to set this during testing so that when tests fail I can look at the trace. 
The reason I made it a feature is that the code itself is in a library, but at the same time it seems more appropriate to have the consuming crate decide whether or not to perform the trace.

To be clear, an actual consuming crate has no problem setting the feature and getting corresponding results. It is just for testing the library crate itself that it doesn't seem possible at the moment.

**Comment by sanmai-NL** [comment]
@alexcrichton:

As of yet, the Cargo panic is gone.

```
  cargo --version
cargo 0.19.0-nightly (fa7584c14 2017-04-26)
```

The output is now as follows:

Excerpt from `Cargo.toml`
```
[features]
default = [] # "clippy"

[profile.test]
default = ["clippy"]
```

```sh
cargo test --color=always --package my_app --bin my_app test_1 -- --nocapture
warning: unused manifest key: profile.test.default
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running target/debug/deps/structure_web_server-3dc5f86e9b4fb259

running 1 test
test tests::test_1 ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
```

You wrote this isn't possible. It appears there are no immediate objections to adding such a feature. Could you explain how fundamental the limitation is? And, if possible, sketch a solution? Just so a PR becomes more likely to happen.

**Comment by alexcrichton** [comment]
Ah I believe this was fixed in https://github.com/rust-lang/cargo/pull/3921 as https://github.com/rust-lang/cargo/issues/3902 was a dupe of this issue. I guess I was wrong before!

**Comment by sanmai-NL** [comment]
@alexcrichton: the crashes have been resolved yes, but the original feature request still stands AFAIK. Reopen?

**Comment by alexcrichton** [comment]
Oh sorry, missed that aspect.

**Comment by sunjay** [comment]
Is there any way to add features per profile now? That is, if I want to enable certain features during tests, can I do that yet?

**Comment by alexcrichton** [comment]
@sunjay unfortunately no, you'll still need to pass `--features` on the command line

**Comment by sunjay** [comment]
Is this change something that would require an RFC or could it just be implemented? It sounds like being able to specify features in the profile configuration is something that could benefit a lot of people.

**Comment by alexcrichton** [comment]
@sunjay I think due to the subtle possible ramifications of auto-enabling features I'd personally want to see an RFC, but it may be worth canvassing others from @rust-lang/cargo first as well

**Comment by sunjay** [comment]
I would love to work on an RFC for this. Very interested to hear what the cargo team thinks.

Right now, I'm having to constantly remember to pass features in when I run cargo test. I would really prefer to not have to do that. It's an extra point of friction for anyone who uses my code. 

Though it would be great to even just be able to configure features in tests, I would really like to see cargo support a features/default features option in each of the [profile.*] sections. My RFC would be proposing something like that.

**Comment by matklad** [comment]
I think we have an RFC by @withoutboats for this already: https://github.com/rust-lang/rfcs/pull/1956, which is postponed because of profiles.

**Comment by sunjay** [comment]
@matklad Ah yup that's exactly what I would have proposed. Thanks for looking that up! I read the discussion and it seems like that's pending on some work with how multiple profiles are applied together (e.g. `test --release`). It seems that the current behavior is more complicated than what it looks like at first glance.

Is there any way to do this in a backwards compatible way? It would be great if there was some way to just add another configuration option that is in effect the same as the command line argument while keeping everything the same as it is now. Would that make it too hard to change to something better (like profiles) later? 

The main issue I'd like to solve is that right now, introducing features into your codebase makes it so that you have to pass features in as command line arguments every time you run a certain commands. If these aren't default features that you need for all builds, it's not possible to simply configure the ones you want in Cargo.toml. This makes it harder for people new to your codebase or to cargo set things up. It is very repetitive and easy to forget. Is there an easier way to just address that issue?

**Comment by brson** [comment]
I've also run into the need to have features automatically enabled in test builds. In my case I want the python bindings "extension-module" feature on for production builds, but off for test builds so that I can link directly to python during testing.

**Comment by Object905** [comment]
Also it's very useful for `#![no_std]` ergonomics. Since you need to test with libstd linked, otherwise you get 
`error: no #[default_lib_allocator] found but one is required; is libstd not linked?`. So you constantly running `cargo test --features std`

**Comment by ebfull** [comment]
Note that you *can* do this with `required-features` if you specify your test targets explicitly, aka:

```
[[test]]
name = "whatever"
path = "tests/whatever.rs"
required-features = ["mycoolfeature"]
```

**Comment by RReverser** [comment]
@ebfull That only skips building/running tests if feature is not enabled instead of force-enabling the feature during build.

**Comment by synek317** [comment]
Some ugly workaround, at least for some cases, is to extend all your #[cfg(feature = ...)] with `test` (e.g. #[cfg(any(test, feature = "..."))] and copy all optional dependencies in [dev-dependencies] and remove `optional = true`.

It works nice for me but for sure there are cases where it is not enough.

**Comment by nixpulvis** [comment]
Yea I feel like an RFC here would make sense if one doesn't already exist. I'm using rust again, and finding myself really wanting something like `cargo test --feature-matrix` for example.

**Comment by Boiethios** [comment]
> Some ugly workaround, at least for some cases, is to extend all your `#[cfg(feature = ...)]` with `test` (e.g. `#[cfg(any(test, feature = "..."))]` and copy all optional dependencies in [dev-dependencies] and remove `optional = true`.
> 
> It works nice for me but for sure there are cases where it is not enough.

I like your trick, but unfortunately, it does not work for the documentation tests (not sure why, though).

**Comment by isavegas** [comment]
Having something akin to `#[features("chrono", "alloc")]` and so on to set a specific feature set for a test module (with `#[features()]` to use no features) would be awesome! Of course, having this functionality available elsewhere is probably not a great idea.

Of course, this might not even be feasible without rewriting the core of the testing framework, but perhaps it could be a nice quality of life feature to look at down the line.

**Comment by cormacrelf** [comment]
If you have a workspace going, a better workaround is to create a `test_utils` crate. You can make the feature activate automatically for an integration test.

```toml
# test_utils crate
[dependencies]
main = { path = "../main", features = ["test"] }
```

```rust
pub use main as main_test;
// or, may as well use this crate to compile the integration testing infrastructure
pub fn exported(file: &str) { main::whatever(); assert_eq!(1, 1); }
```

```toml
// main crate
[features]
test = []

[dev-dependencies]
datatest = "..."
test_utils = { path = "../test_utils" }
```

```rust
// lib.rs
cfg_if::cfg_if! {
    if #[cfg(feature = "test")] {
        fn whatever() { do_something_else(); }
    } else {
        fn whatever() {}
    }
}
```

```rust
# tests/integration.rs
use test_utils::main_test as main;
use test_utils::exported;

#[datatest::files( ... )]
fn integration(file: &str) {
   exported(file);
}
```

Lotta moving parts but I think I wrote it all right.

**Comment by Diggsey** [comment]
What's the current state on this, has there been another RFC?

**Comment by Nashenas88** [comment]
@cormacrelf unfortunately that doesn't work for my case, where I have multiple features I'd like to test. My current workaround is to use a makefile to keep track of different test crates, and each of those depends on the main crate with specific features turned on. With workspaces, you can't mix features. Would really love to see some  progress here, and I'd be willing to help with the RFC if help is needed.

**Comment by ChaseElectr** [comment]
We are working on a really big project and use a lot of protobuf structs. Now we want enable Serialize and Deserialize on those struct only under test situations, but found control through features may be our only choice:
First we have a lot of structs define like this
```
// crate a, which dependens on crate b
#[cfg_attr(feature = "with-serde", Serialize, Deserialize)]
struct Foo {
    bar: Bar
}

// crate b
#[cfg_attr(feature = "with-serde", Serialize, Deserialize)]
struct Bar {
    b: i32
}
```
Then we want write tests for crate a
```
#[cfg(test)]
mod tests {
    #[test]
    fn my_test() {
         let foo: Foo = serde_json::from_str("xxxxx");
         balabala.....
    }
}
```
We have to add `#[cfg(feature = "with-serde")]` for `my_test` or for `mod tests`, or the clippy will complain about not implement `Serialize`/`Deserialize` for `Foo`.

We alse have tried with using `#[cfg_attr(test, Serialize, Deserialize)]` instead of `#[cfg_attr(feature = "with-serde", Serialize, Deserialize)]`, but found that when we run test for crate a, crate b won't have `cfg_attr(test)` to be true, which will lead `Serialize`/`Deserialize` not implement for Bar.
Another choice is use other config attribute instead of `test`, such as `#[cfg_attr(with_serde, Serialize, Deserialize)]`, but this needs additional effort for both writing tests and running tests.

If we can have default features for test, it seems is the simplest way to solve this problem😋

**Comment by Urhengulas** [comment]
Hi :wave:,

I have a bit of a simpler case, where I only want to activate one feature (`testing`), which serves the test infrastructure for all tests.

The simplest workaround I could figure out is to add the library itself with the desired feature to the dev-dependencies in `Cargo.toml`:
```toml
[dev-dependencies]
my_lib = { path = ".", features = ["testing"] }

[features]
testing = []
```

This is obviously not as fine-grained as per-profile-features, but maybe helps some folks 😁 

Cheers

**Comment by DesmondWillowbrook** [comment]
There is currently a work-around for this. ([Source](https://github.com/rust-lang/cargo/pull/3921#issuecomment-297421636))
```toml
[dev-dependencies]
self-package-name= { path = ".", features = ["desired_feature"] }
```
where `self-package-name` is the name of the package in the `Cargo.toml` file.

**Comment by sander2** [comment]
> There is currently a work-around for this. ([Source](https://github.com/rust-lang/cargo/pull/3921#issuecomment-297421636))

@DesmondWillowbrook This seems to activate `desired_feature` even during non-dev build, for some reason.

**Comment by DesmondWillowbrook** [comment]
You're right! This _is_ weird.

**Comment by Owez** [comment]
> There is currently a work-around for this. ([Source](https://github.com/rust-lang/cargo/pull/3921#issuecomment-297421636))
> 
> ```toml
> [dev-dependencies]
> self-package-name= { path = ".", features = ["desired_feature"] }
> ```
> 
> where `self-package-name` is the name of the package in the `Cargo.toml` file.
> 
> [Sample repo for the same](https://github.com/DesmondWillowbrook/cargo-optional-features-for-testing-and-examples)

Unfortunately this doesn't seem to work on an edge case in a `no_std` environment where your crate allows `no_std` and has a feature which enables `std`. Here's an example of where the issue can occur:

```none
error[E0152]: found duplicate lang item `panic_impl`
  --> package/src/main.rs:15:1
   |
15 | fn panic(_info: &PanicInfo) -> ! {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the lang item is first defined in crate `std` (which `sprawl_format` depends on)
   = note: first definition in `std` loaded from /home/owen/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-3d933644112ba794.so, /home/owen/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-3d933644112ba794.rlib
   = note: second definition in the local crate (`sprawl_package`)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0152`.
error: could not compile `sprawl-package`

To learn more, run the command again with --verbose.
```

I assume this is because its importing `std` for the developer dependencies section, and this function is trying to overwrite a now-included package.

**Comment by Owez** [comment]
The workaround for this workaround seems to be

```none
#[cfg(not(test))]
// std overriding code here
```

In cases where std is acceptable if you still want to test :smile:

**Comment by daxpedda** [comment]
> This seems to activate `desired_feature` even during non-dev build, for some reason.

> Unfortunately this doesn't seem to work on an edge case in a `no_std` environment where your crate allows `no_std` and has a feature which enables `std`.

> I assume this is because its importing `std` for the developer dependencies section, and this function is trying to overwrite a now-included stdlib function.

> EDIT: This may actually be a bug with workspacing? if I delete my workspace file and use them separately this works fine. The other option is that I'm doing something wrong :smile:

Indeed the issue is simply because of the old bug of features from dependencies and dev-dependencies being unified, see #4866. This is fixed by using the new resolver.

**Comment by H1365** [comment]
God

**Comment by d-e-s-o** [comment]
> There is currently a work-around for this. ([Source](https://github.com/rust-lang/cargo/pull/3921#issuecomment-297421636))
> 
> ```toml
> [dev-dependencies]
> self-package-name= { path = ".", features = ["desired_feature"] }
> ```
> 
> where `self-package-name` is the name of the package in the `Cargo.toml` file.
> 
> [Sample repo for the same](https://github.com/DesmondWillowbrook/cargo-optional-features-for-testing-and-examples)

This may be another thing worth pointing out with this approach: `cargo` seems to strip this specific reference to `self-package-name` from `dev-dependencies` when `package`ing and `publish`ing. Meaning if you require `desired_feature` for, say, testing the package, tests may not work as expected on the published package unless you manually enable this feature. Typically that's not a problem, but it was unexpected to me and one can certainly construct cases where it matters.

**Comment by mattmccutchen-amazon** [comment]
> ```toml
> [dev-dependencies]
> self-package-name= { path = ".", features = ["desired_feature"] }
> ```

I tried this, and it worked like a charm except that when I run `cargo test --no-default-features`, the default features of my package are still enabled.  My theory is that since the explicit development dependency on `self-package-name` does not specify `default-features = false`, it activates the default features of the target package just like any other dependency would, but in this case, the target package happens to be the current package.  So even if I tell Cargo not to activate the default features, they still get activated by [Cargo's feature unification rules](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification).  If I change the workaround to:

```toml
[dev-dependencies]
self-package-name = { path = ".", default-features = false, features = ["desired_feature"] }
```

then I get the behavior I want: plain `cargo test` activates the default features and `cargo test --no-default-features` doesn't.  So I posit that the above is a superior version of the workaround in general, but if you don't use `cargo test --no-default-features`, it's fine to stick with the simpler version.

**Comment by epage** [comment]
Allowing build-targets to force features to be on was proposed in rust-lang/rfcs#3374 but that was closed due to the challenges with feature unification and the ability to ensure a code base works with all combinations of publicly available features

**Comment by epage** [comment]
I'm going to propose to the cargo team that we close this

We have the following workarounds
- According to the comments, having a dev-dependency on yourself (so long as `package.resolver="2"` is set) will work
- `test.required-features` exists
- `#[cfg()]` on the relevant tests/mods

We also have #1982 tracking build-target-specific dependencies.   rust-lang/rfcs#3374  was an offshoot of that which was closed due to challenges with feature unification.

**Comment by Qix-** [comment]
> `test.required-features` exists

As far as I'm understanding that only enables the tests if the features are all available, not the other way around. The `dev-dependencies` thing also seems like a hack.

**Comment by epage** [comment]
> As far as I'm understanding that only enables the tests if the features are all available, not the other way around.

What would help is to say what use case is not well served by `required-features`.  

Something to also keep in mind is that this feature was in a way covered by  rust-lang/rfcs#3374 but it was abandoned.  Feature unification is already a really difficult problem for users and it made it even worse.

**Comment by Qix-** [comment]
Because running _all_ tests is important, and not all tests need _every_ feature (or even work with every feature). The unfortunate reality is that "features should be additive" isn't possible in every single case, so the assumption that `--all-features` work in all cases isn't sound. Thus, in order to run every single test possible in a crate, in cases where some features are incompatible with others, then there must be different feature sets enabled in different cases in order to run 100% of the tests.

As it stands, that means multiple invocations, and even then, there's a high chance you miss a test suite because a feature config was changed somewhere but whatever script you use to run each `cargo test` invocation wasn't also updated, and there's no way to error or fail the CI/CD process when that occurs because there's no way to detect such a case.

**Comment by epage** [comment]
I think i'm missing something because it sounds like you are saying that enabling all features is a problem, meaning you need to enable different sets of features.  To me, that sounds like a test shouldn't be enabling features automatically but should instead be left up to the caller.  On top of this, having tests enable some features means that you can't test without them and it is easy to break feature interactions that you don't test.

Some ways of dealing with testing
- When its not too much, [cargo hack](https://crates.io/crates/cargo-hack) can automatically enumerate combinations of features
- Having a build orchestrator capture specific sets of features used and test them (in `clap` I use make for this)
  - #10882 would allow encoding these feature sets as features

**Comment by ilyagr** [comment]
In case of `jj`, the tests use a couple of [binaries](https://github.com/martinvonz/jj/blob/db538bd0f4f9be66a5e984f856888776c9408f9b/cli/Cargo.toml#L19-L27) for a "fake editor" (for example). We need some way to not have these binaries installed when the user runs `cargo install jj-cli`. The lack of fake editor in the binary releases also caused `cargo binstall` to refuse to install `jj-cli`.

In https://github.com/martinvonz/jj/pull/2277, I used the dev-dependencies hack to acheive this. Unfortunately, that doesn't work well with rust-analyzer, see e.g. https://github.com/rust-lang/rust-analyzer/issues/14167 . Other than that, it'd work.

For my purpose, I don't need any conditional compilation within a crate, only a way to prevent generating some binaries outside of tests.

**Comment by epage** [comment]
Thanks @ilyagr for the clear use case!

Access to test binaries is a sore point right now I can see how this helps in that case.  Another feature that could help is [artifact dependencies](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#artifact-dependencies)

**Comment by martinling** [comment]
Here's another use case from our GUI application [Packetry](https://github.com/greatscottgadgets/packetry/).

Although our codebase does include some unit tests written in the usual way with `#[cfg(test)]` and `#[test]`, the vast majority of testing is carried out by a specific test called `test_replay`. This test replays a large amount of input data and user actions into the UI code, and checks the results.

That test has to be declared in `Cargo.toml` with `harness = false`, because it calls into GTK, and on macOS, GTK can only be called from a program's original main thread. The standard cargo test harness does not run tests in the main thread, even if you run `cargo test` with `-j 1`. So we have to write the test as a standalone binary without the standard harness, or it won't work on macOS.

To achieve the testing we want, we have to add a bunch of logging output to our normal UI code. We don't ever want that logging code at normal runtime. It's only there for testing. But we can't gate it behind `#[cfg(test)]`, because that isn't enabled when building our separate test binary: it only applies for tests specified in the standard way. So we have to gate the logging code behind a feature instead, which we call `test-ui-replay`.

Most of the above details aren't relevant: what matters is just that we have a custom test, not using the standard harness, which requires some feature, and which we want to always be run.

In `Cargo.toml`, we use `required-features` to specify that the `test_replay` test requires the `test-ui-replay` feature. But we still have to specifically request that feature, every time we test: `cargo test --features=test-ui-replay`

What we want is to be able to just run `cargo test`, and have this test always be run. It's absolutely fine that we won't be able to test without that 'feature'. That's exactly what we want. In our case it's not even a feature in the normal sense - it's only there because we can't use `#[cfg(test)]`, which in turn is because we're not using the standard test harness for that test.

I've just tried, and the dev-dependency hack works for us. But that's not documented anywhere other than this thread, and the discussion in #9518 says that this isn't really something that should be allowed, and the only reason it's not being made an error yet is that it allows working round this issue. That doesn't sound like something we should be relying on.

I am opposed to closing this issue without either a proper solution, or another issue left open tracking what's needed.

The proposal to close suggests:
- the self-dev-dependency hack: as above, this seems to only work by accident, and is not documented as supported.
- `test.required-features`: we're already using this, it doesn't solve our problem.
- `#[cfg()]` on the relevant tests/mods: we're already using this, it doesn't solve our problem.
- #1982: this only covers build-target-specific *dependencies*, not features required in the current package.
- #3374: this RFC is closed, and there's no other current proposal.

**Comment by epage** [comment]
> Most of the above details aren't relevant: what matters is just that we have a custom test, not using the standard harness, which requires some feature, and which we want to always be run.

Actually it is relevant, particularly the combination of (1) tests that need access to private APIs or behaviors and (2) not being able to write these as unit tests.

This kind of information is the core of whats needed to decide how important a use case is and how to handle it.

Is the logged information access through a feature-gated API or is it exposed through stdio / GUI?

**Comment by epage** [comment]
@ilyagr another potential workaround is to have the test bins auto-discovered b
but to ensure they aren't included in the published artifact (`.crate`) via `package.include` or `package.exclude`.

**Comment by ilyagr** [comment]
@epage 

> another potential workaround is to have the test bins auto-discovered b but to ensure they aren't included in the published artifact (`.crate`) via `package.include` or `package.exclude`.

Could you elaborate? I'm far from an expert on `Cargo.toml`. I'm not sure what "auto-discovered" means.

Naively, I was wondering whether there could be a dev-dependency on a separate crate for the fake binaries, but I'm not sure whether the tests will be able to find the binaries from another crate.

**Comment by martinling** [comment]
> Actually it is relevant, particularly the combination of (1) tests that need access to private APIs or behaviors and (2) not being able to write these as unit tests (with enough context to understand the "why")

I take your point; I was just trying to pre-emptively avoid the discussion getting dragged into the weeds of why GTK/macOS has that particular annoying limitation, when there are any number of other reasons why someone might want to have tests that don't use the standard test harness.

> Is the logged information access through a feature-gated API or is it exposed through stdio / GUI?

The test creates a UI in the same way the normal program would, but the feature causes the UI to be created off-screen, with logging code. The test then uses feature-gated APIs on the UI to set where to direct the logging, and also to inject a sequence of pre-recorded user actions into the UI. The log output is then compared against a reference file, and if the two match, the test passes.

As I understand it, the reason that `#[cfg(test)]` doesn't work for us is that the UI code is in a library target, and that library target is built without `cfg(test)` set, because a standalone test is presumed to be an integration test that only uses public APIs. In fact, we don't have any truly public API at all - we don't publish a library crate. The library target only exists so that we can share UI code between the main program and the replay test.

**Comment by martinling** [comment]
For our particular use case, if there were a way to tag a unit test as "must be run in the main thread of a process", and have Cargo honour that somehow (either by executing from its main thread or spawning a new process), then that would be a better solution for us. Originally `test_replay` _was_ just a unit test in `main.rs`; it's only because that wouldn't work on macOS that I had to break up `main.rs` into `lib.rs`, `ui.rs`, and `test_replay.rs` so that it could run as a standalone test.

**Comment by martinling** [comment]
> Put most of your code in a lib and have another package depend on it, activating the feature that is needed, and have the tests live there.

I have just verified that yes, it is possible to make `cargo test` work as desired by further splitting what was originally a simple, single-binary project into three different packages, with four `Cargo.toml` files. But what was the point of this suggestion?

Yes, it is possible to work around the lack of the feature that people are asking for in this issue. That doesn't make the feature request invalid. The complexity of the workarounds you're suggesting illustrate exactly why it would be helpful.

> Adding accessibility support (which would likely be a lot of work) as that tends to have enough overlap for testing

Even with a lot of work, I do not believe that it would be either practical or performant for us to implement the same test through the GTK accessibility APIs. We're not just clicking on some buttons and looking for dialog messages. We're dealing with a list model with what may be millions of rows that the UI loads from on demand. And the test also needs to control the inflow of packets into the decoder engine so that it can verify how descriptions of items are updated by new packets. There's no way to do all that through the normal application UI, we have to hook into it specially, and that requires gating some code.

But even if we could, the fact that this may a suitable solution for _some_ GUI applications doesn't invalidate the use cases where it *isn't*.

Why is there so much desire to close this? I don't really see anyone asking for anything complicated here. A solution could be as simple as a `Cargo.toml` key that makes `cargo test` behave as if you'd typed `cargo test --features=foo,bar`. As far as I can see, it doesn't need to affect how feature unification works, or any other Cargo internals. It's just about setting some defaults per-profile on the frontend side.

**Comment by epage** [comment]
> I have just verified that yes, it is possible to make `cargo test` work as desired by further splitting what was originally a simple, single-binary project into three different packages, with four `Cargo.toml` files. But what was the point of this suggestion?
> 
> Yes, it is possible to work around the lack of the feature that people are asking for in this issue. That doesn't make the feature request invalid. The complexity of the workarounds you're suggesting illustrate exactly why it would be helpful.

I would assume you could get away with just two packages and two `Cargo.toml` files (workspace + lib + main bin, test bin)

As for the point, it was to offer another alternative solution for people running into it.  It might work for some people; it might not work for others.  It does offer the benefit of having the minimal affect on feature activations which is likely needed in some other cases.

I do feel it important to call out that what is important in an issue is the use cases.  We shouldn't get attached to any given solution.

> Why is there so much desire to close this? I don't really see anyone asking for anything complicated here. A solution could be as simple as a Cargo.toml key that makes cargo test behave as if you'd typed cargo test --features=foo,bar. As far as I can see, it doesn't need to affect how feature unification works, or any other Cargo internals. It's just about setting some defaults per-profile on the frontend side.

I think there might be a misunderstanding.  None of this is "arguing back to force through a closure".  I am trying to understand use cases and explore the problem space.  There are valid workflow problems people are having and they weren't all communicated earlier.  The question is whether this is still the right solution.  That, I'm still not sure on.

And yes, we can add a lot of little features like this but
- I've found the more features you have, the less value you get out of them because you overwhelm users and they can't find what they need to solve problems.  Its best to look for patterns in case you can find wider ways of solving problems.  There might even be times when its appropriate to say "this is too narrow of case for built-in support so long as there are valid workarounds"
- We need to consider the whole workflow and how the feature is used to ensure we are helping drive the right behavior, while still allowing the exception cases.  Otherwise, a "simple" issue like this could turn into 10 more issues that want to extend whatever solution we come up with in 100 different ways because it doesn't quite work for them.

**Comment by martinling** [comment]
> I think there might be a misunderstanding. None of this is "arguing back to force through a closure". I am trying to understand use cases and explore the problem space.

That intent would have been clearer had you not first entered the discussion with a proposal to close, and then proceeded to respond to use cases by listing workarounds.

But anyway. In our specific case, as I already said, I don't think this feature is actually the _best_ solution to our problem. What we have is really a unit test, we're just lacking a way to specify that this unit test must be executed on the main thread of a program. I guess I should go raise that as a feature request for libtest.

**Comment by weihanglo** [comment]
> A solution could be as simple as a `Cargo.toml` key that makes `cargo test` behave as if you'd typed `cargo test --features=foo,bar`. As far as I can see, it doesn't need to affect how feature unification works, or any other Cargo internals.

Are you proposing something like #4942? It doesn't seem too hard in terms of supporting that. However but I've outlined [some unresolved questions](https://github.com/rust-lang/cargo/issues/4942#issuecomment-1762853408) around that needing to be addressed.

If we look into RFC 3374 a bit deeper, there is a paragraph calling out [some difficulties and open questions](https://github.com/ehuss/rfcs/blob/cargo-target-features/text/3374-cargo-target-features.md#change-the-required-features-behavior) on the implementation side. I know it looks not too hard but things are under the surface 😞.

Anyway, thank you for bringing more info on the table.

**Comment by Teebor-Choka** [comment]
What is the best strategy to test library code behind features in integration tests, if the features are not enabled by default?

**Comment by hpux735** [comment]
> What is the best strategy to test library code behind features in integration tests, if the features are not enabled by default?

I'd like to know this too.  I have a library that supports std and no-std.  I don't know how to be able to test the different "personalities" of the library.

**Comment by ctz** [comment]
> I'm going to propose to the cargo team that we close this

As the original reporter -- still checking in after 9.5 years -- I agree.

In closing, crate features and the UX around them is and remains very poor and a significant weak point of cargo. I now avoid their use to the maximum extent possible in my projects, just so I don't have to think about how awful and underbaked they are.

**Comment by marstaik** [comment]
Yeah, the user experience on this is kind of terrible.

Lets say you have tests that need some production-invalid (test only) features, like `mock-allocator`, or you want to gate certain features for performance, like `screenshots` which require a vulkan instance (in my case). `mock-allocator` cannot work with `screenshots`, so you cant just make a script wrapper/command that activates them both.

You can tag those tests with cfg, fine. But they you have to have a bunch of test scripts or remember to pass those features to run your tests, and have multiple of them. This also leads to running the same tests *multiple times*, as when you enable a feature, it doesn't prevent the others without the feature from running. So the unit tests run once with no features, again on feature `mock-allocator`, and again on feature `test-screenshots`.

It's fairly clear that if a test has a feature gate on it, that test *needs* that feature gate, or lack of it, and those tests could be grouped into different combined binaries automatically.

the `cargo test` command by itself could group tests by their binary-run-group, and execute.
```
[no features]
standard-unit-test

[features(mock-allocator)]
mock-allocator-test-a
mock-allocator-test-b

[features(mock-allocator & mutex-injection)]
mock-allocator-test-c-with-mutex-injection

[features(screenshots)]
screenshot-a
screnshot-b
```
